core/shared/src/main/scala/scalaz/zio/stream/Chunk.scala

```diff
@@ -0,0 +1,874 @@
+       else Chunk.Arr(dest))
+  }
+
+  /**
+   * Takes the first `n` elements of the chunk.
+   */
+  final def take(n: Int): Chunk[A] =
+    if (n <= 0) Chunk.Empty
+    else if (n >= length) this
+    else
+      self match {
+        case Chunk.Empty => Chunk.Empty
+        case Chunk.Slice(c, o, l) =>
+          if (n >= l) this
+          else Chunk.Slice(c, o, n)
+        case c @ Chunk.Singleton(_) => c
+        case _                      => Chunk.Slice(self, 0, n)
+      }
+
+  /**
+   * Takes all elements so long as the predicate returns true.
+   */
+  def takeWhile(f: A => Boolean): Chunk[A] = {
+    val len = self.length
+
+    var i = 0
+    while (i < len && f(self(i))) {
+      i += 1
+    }
+
+    if (i == 0) Chunk.Empty
+    else if (i == len) this
+    else Chunk.Slice(this, 0, i)
+  }
+
+  /**
+   * Converts the chunk into an array.
+   */
+  def toArray[A1 >: A]: Array[A1] = {
+    implicit val A1: ClassTag[A1] = Chunk.classTagOf(self)
+
+    val dest = Array.ofDim[A1](self.length)
+
+    self.toArray(0, dest)
+
+    dest
+  }
+
+  def toSeq: Seq[A] = {
+    val seqBuilder = Seq.newBuilder[A]
+    var i          = 0
+    val len        = length
+    seqBuilder.sizeHint(len)
+    while (i < len) {
+      seqBuilder += apply(i)
+      i += 1
+    }
+    seqBuilder.result()
+  }
+
+  override def toString: String =
+    toArray.mkString(s"${self.getClass.getSimpleName}(", ",", ")")
+
+  /**
+   * Effectfully traverses the elements of this chunk.
+   */
+  final def traverse[E, B](f: A => IO[E, B]): IO[E, Chunk[B]] = {
+    val len                    = self.length
+    var array: IO[E, Array[B]] = IO.now(null.asInstanceOf[Array[B]])
+    var i                      = 0
+
+    while (i < len) {
+      array = array.seqWith(f(self.apply(i))) { (array, b) =>
+        val array2 = if (array == null) {
+          implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+          Array.ofDim[B](len)
+        } else array
+
+        array2(i) = b
+        array2
+      }
+
+      i += 1
+    }
+
+    array.map(
+      array =>
+        if (array == null) Chunk.empty
+        else Chunk.fromArray(array)
+    )
+  }
+
+  /**
+   * Effectfully traverses the elements of this chunk purely for the effects.
+   */
+  final def traverse_[E](f: A => IO[E, Unit]): IO[E, Unit] = {
+    val len             = self.length
+    var io: IO[E, Unit] = IO.unit
+    var i               = 0
+
+    while (i < len) {
+      io = io *> f(self(i))
+      i += 1
+    }
+
+    io
+  }
+
+  /**
+   * Returns two splits of this chunk at the specified index.
+   */
+  final def splitAt(n: Int): (Chunk[A], Chunk[A]) =
+    (take(n), drop(n))
+
+  /**
+   * Zips this chunk with the specified chunk using the specified combiner.
+   */
+  final def zipWith[@specialized B, @specialized C](that: Chunk[B])(f: (A, B) => C): Chunk[C] = {
+    val size = self.length.min(that.length)
+
+    if (size == 0) Chunk.empty
+    else {
+      var dest = null.asInstanceOf[Array[C]]
```

Comment:
At this point you now that `size > 0` so you can as well allocate the array here

Reply:
Well, the problem is that you need the first element to allocate the array, as you need the classtag. This means we need to replicate the loop logic for the first element, which I'm not sure is a big win in terms of performance and maintainability.

Let's leave this to benchmarking and later PRs?